[LiHaixin000] iP

LiHaixin000 opened the pull request:
# Arts Application

> “Your mind is for having ideas, not holding them.” – David Allen

**Arts** is a task management application that helps you organize your tasks efficiently. It's:

- Text-based
- Easy to learn
- **FAST** and _SUPER FAST_ to use

All you need to do is:

1. Download it from [here](https://github.com/LiHaixin000/ip).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you! 😊

And it is **FREE**!

## Features

- [T][ ] Managing Todos
- [D][ ] Managing deadlines
- [E][ ] Managing events
- [ ][ ] Reminders (coming soon)

### Code Example

Here's a simple Java code example:

```java
package arts;

import arts.util.Parser;
import arts.util.Storage;
import arts.util.Ui;
import arts.task.TaskList;
import arts.enums.CommandType;
import arts.command.AddDeadlineCommand;
import arts.command.AddEventCommand;
import arts.command.MarkCommand;
import arts.command.UnmarkCommand;
import arts.command.DeleteCommand;
import arts.command.AddTodoCommand;
import arts.command.FindCommand;

import java.time.format.DateTimeFormatter;

/**
 * The Arts class is the main entry point for the Arts application.
 * It manages the initialization and execution of the application, handling user commands
 * and coordinating between different components like storage, UI, and task management.
 */
public class Arts {
    private static final DateTimeFormatter[] INPUT_FORMATTERS = new DateTimeFormatter[]{
            DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"),
            DateTimeFormatter.ofPattern("d/M/yyyy HHmm")
    };

    private final Storage storage;
    private final TaskList tasks;
    private final Ui ui;
    private final Parser parser;

    /**
     * Constructs an Arts object with the specified file path for task storage.
     * Initializes the UI, storage, parser, and loads existing tasks.
     *
     * @param filePath The path of the file where tasks are stored.
     */
    public Arts(String filePath) {
        ui = new Ui();
        storage = new Storage(filePath);
        parser = new Parser();
        TaskList tempTasks;
        try {
            tempTasks = new TaskList(storage.load());
        } catch (ArtsException e) {
            ui.showError(e.getMessage());
            tempTasks = new TaskList();
        }
        tasks = tempTasks;
    }

    /**
     * Runs the Arts application, processing user commands in a loop until exit.
     */
    public void run() {
        ui.showWelcome();
        boolean isExit = false;
        while (!isExit) {
            try {
                String input = ui.readCommand();
                CommandType command = parser.parseCommand(input);
                String[] parts = parser.parseArguments(input);

                switch (command) {
                    case BYE:
                        ui.showGoodbye();
                        isExit = true;
                        break;
                    case LIST:
                        ui.showMessage(listTasks());
                        break;
                    case MARK:
                        new MarkCommand(tasks, storage, ui, parts[1]).execute();
                        break;
                    case UNMARK:
                        new UnmarkCommand(tasks, storage, ui, parts[1]).execute();
                        break;
                    case DELETE:
                        new DeleteCommand(tasks, storage, ui, parts[1]).execute();
                        break;
                    case TODO:
                        new AddTodoCommand(tasks, storage, ui, parts[1]).execute();
                        break;
                    case DEADLINE:
                        new AddDeadlineCommand(tasks, storage, ui, parts[1], INPUT_FORMATTERS).execute();
                        break;
                    case EVENT:
                        new AddEventCommand(tasks, storage, ui, parts[1], INPUT_FORMATTERS).execute();
                        break;
                    case FIND:
                        new FindCommand(tasks, ui, parts[1]).execute();
                        break;
                    default:
                        throw new ArtsException("I'm sorry, but I don't know what that means.");
                }
            } catch (ArtsException e) {
                ui.showError(e.getMessage());
            } catch (Exception e) {
                ui.showError("An unexpected error occurred: " + e.getMessage());
            }
        }
    }

    /**
     * Lists all tasks currently in the task list.
     *
     * @return A string representation of all tasks.
     */
    private String listTasks() {
        if (tasks.isEmpty()) {
            return "No tasks yet! Why not add some?";
        } else {
            StringBuilder sb = new StringBuilder("Here are the tasks in your list:\n");
            for (int i = 0; i < tasks.size(); i++) {
                sb.append((i + 1)).append(". ").append(tasks.getTask(i)).append("\n");
            }
            return sb.toString();
        }
    }

    /**
     * Main method to start the Arts application.
     *
     * @param args Command line arguments.
     */
    public static void main(String[] args) {
        new Arts("./data/tasks.txt").run();
    }
}

```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

LiHaixin000 committed:
Add .class Files to .gitignore

LiHaixin000 committed:
Rename, Greet, Exit

LiHaixin000 committed:
Improve the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye

LiHaixin000 committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested

LiHaixin000 committed:
Added task class

LiHaixin000 committed:
Add the ability to mark tasks as done and undone

LiHaixin000 committed:
Added Deadline, Event and Todo classes

LiHaixin000 committed:
Implemented behaviours for ToDos, Events, Deadlines in Arts class

LiHaixin000 committed:
implemented automated text UI testing

LiHaixin000 committed:
Added trim to inputs

LiHaixin000 committed:
Added ArtsException class and added try-catch error handling to Arts

LiHaixin000 committed:
Add support for deleting tasks from the list

LiHaixin000 committed:
Added enum Command class

LiHaixin000 committed:
Use enums in Arts class

LiHaixin000 committed:
Updated input and expected for text UI test to include new features

LiHaixin000 committed:
Implement task saving and loading functionality

LiHaixin000 committed:
Merge branch-Level-7 implementing Level 7

LiHaixin000 committed:
Implement save tasks to file

LiHaixin000 committed:
Replace string with dates and times

LiHaixin000 committed:
Merge branch-Level-8 implementing Level 8

LiHaixin000 committed:
Implement OOP to extract out closely related code as classes

LiHaixin000 committed:
Merge A-MoreOOP implement More OOP

LiHaixin000 committed:
Divide classes into packages

LiHaixin000 committed:
Merge branch 'add-gradle-support'

LiHaixin000 committed:
Add JUnit support

LiHaixin000 committed:
Modify build.gradle

LiHaixin000 committed:
Implement JavaDoc comments

LiHaixin000 committed:
Apply coding standards

LiHaixin000 committed:
Implement find function

LiHaixin000 committed:
Merge branch 'branch-A-CodingStandard' to apply coding standards updates.

LiHaixin000 committed:
Resolve conflicts from merging

GavinSin commented on src/main/java/arts/Arts.java:
```diff
@@ -0,0 +1,131 @@
+                String[] parts = parser.parseArguments(input);
+
+                switch (command) {
+                    case BYE:
```
There should be no indentation for case clauses according to the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html) mentioned.

You can tweak your Intellij IDE to ensure future switch case follow no indentation [here](https://se-education.org/guides/tutorials/intellijCodeStyle.html#tweak-switch-case-style).

GavinSin commented on src/main/java/arts/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,82 @@
+import arts.util.Ui;
+import arts.task.Deadline;
+import arts.ArtsException;
+import java.time.LocalDateTime;
```
Should the java default import statements separated from own import statements?

i.e. 
```
import java.time.LocalDateTime;

import arts.task.TaskList;
``` 
instead of 
```
import java.time.LocalDateTime;
import arts.task.TaskList;
```

GavinSin commented on src/main/java/arts/util/Ui.java:
```diff
@@ -0,0 +1,82 @@
+     * Displays the welcome message and logo to the user.
+     */
+    public void showWelcome() {
+        String logo = "     _    _____  _______  _____  \n"
```
I like that you generated an ASCII logo for your chatbot :) !

GavinSin reviewed (commented):
I like that you have detailed JavaDoc comments for all of your classes and methods, and comprehensive testing. There is only a little coding standard issues I have commented.

In summary:
1. switch case clause no indentation
2. consistent import statements with separation

LiHaixin000 committed:
Add A-CheckStyle increment

LiHaixin000 committed:
Implement basic GUI interface

LiHaixin000 committed:
Improve GUI implementation

LiHaixin000 committed:
Test arts functionalities

LiHaixin000 committed:
Add varargs to my code

YANYAN-w commented on src/main/java/arts/Arts.java:
```diff
@@ -0,0 +1,140 @@
+    private final Ui ui;
+    private final Parser parser;
+
+    /**
+     * Constructs an Arts object with the specified file path for task storage.
+     * Initializes the UI, storage, parser, and loads existing tasks.
+     *
+     * @param filePath The path of the file where tasks are stored.
+     */
```
I like that your header comments for public methods are clear and comprehensive!

YANYAN-w commented on src/main/java/arts/Arts.java:
```diff
@@ -0,0 +1,140 @@
+import arts.util.Storage;
+import arts.util.Ui;
+
+/**
+ * The Arts class is the main entry point for the Arts application.
+ * It manages the initialization and execution of the application, handling user commands
+ * and coordinating between different components like storage, UI, and task management.
+ */
```
I like that your Javadoc comment is descriptive!

YANYAN-w commented on src/main/java/arts/Arts.java:
```diff
@@ -0,0 +1,140 @@
+                return new FindCommand(tasks, parts[1]).execute();
+            default:
+                throw new ArtsException("I'm sorry, but I don't know what that means.");
+            }
```
I like that you follow the coding standard for the switch statement!

YANYAN-w commented on src/main/java/arts/task/Event.java:
```diff
@@ -0,0 +1,53 @@
+    @Override
+    public String toString() {
+        DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern("MMM dd yyyy, h:mm a");
+        return "[E]" + super.toString() + " (from: " + from.format(outputFormatter)
+                + " to: " + to.format(outputFormatter) + ")";
```
I like how you placed a line break in the return statement to improve readability!

LiHaixin000 committed:
Add assertions to check critical assumptions

- Added assertions to ensure input parameters are not null.
- Checked collection sizes to prevent index errors.
- Verified important variables after complex calculations.
- Validated state transitions in stateful components.

These assertions help catch unexpected states during development.

LiHaixin000 committed:
Refactor AddDeadlineCommand to Improve Code Quality

- Minimized the scope of the `date` variable in the `parseDate` method by removing its declaration outside the loop. The variable is now declared within the `try` block, where it is used.
- Extracted the error message in `parseDate` into a constant `DATE_FORMAT_ERROR_MESSAGE` to improve readability and maintainability by avoiding magic strings.
- Added a comment in the empty catch block of `parseDate` to clarify its purpose, which is to continue trying the next formatter if a `DateTimeParseException` occurs.
- Ensured that the happy path is prominent by structuring the code for clarity and handling exceptions as deviations.

These changes enhance the readability, maintainability, and overall quality of the code without altering its logic or functionality.

LiHaixin000 committed:
Refactor AddEventCommand to improve code quality

- Extracted error messages into constants for better readability and maintainability. This helps avoid magic strings and makes it easier to update messages in the future.
- Simplified the date parsing logic by removing the unnecessary variable declaration and directly returning the parsed date. This minimizes variable scope and improves code clarity.
- Used String.format for constructing the return message, which simplifies the string concatenation process and enhances readability.
- Added a comment in the empty catch block to clarify its purpose, ensuring future developers understand why it is intentionally left blank.
- Maintained the same logic and functionality while making the happy path more prominent and improving overall code readability.

These changes aim to enhance the maintainability and clarity of the AddEventCommand class without altering its behavior.

LiHaixin000 committed:
Refactor code to improve code quality

LiHaixin000 committed:
feat: Refactor date parsing using Java Streams in AddDeadlineCommand and AddEventCommand

- Refactored the parseDate method in both AddDeadlineCommand and AddEventCommand to utilize Java Streams.
- Used Streams to iterate over DateTimeFormatter array for more concise and readable code.
- Implemented map, filter, and findFirst operations to attempt parsing dates with multiple formatters.
- Enhanced error handling by throwing ArtsException if no formatter successfully parses the date.

This change improves the code's readability and leverages Java's functional programming capabilities for processing collections.

LiHaixin000 committed:
Merge pull request #1 from LiHaixin000/branch-A-Assertions

Add assertions to check critical assumptions

LiHaixin000 committed:
Merge branch 'master' into branch-A-CodeQuality

LiHaixin000 committed:
Merge pull request from branch-A-CodeQuality to master 

Enhancements:
- Improved code assertions for better error handling and validation.
- Refactored string formatting to use `String.format` for consistency and readability.
- Ensured consistent use of `DateTimeFormatter` for date parsing and formatting.
- Added detailed JavaDoc comments for better code documentation and understanding.

LiHaixin000 committed:
Resolve merge conflict between master and branch-A-CodeQuality

LiHaixin000 committed:
Resolve merge conflicts

LiHaixin000 committed:
Merge branch 'master' into branch-A-Streams

LiHaixin000 committed:
Merge pull request from branch-A-Streams to master

feat: Refactor date parsing using Java Streams in AddDeadlineCommand …

LiHaixin000 committed:
Resolve conflicts in merging

LiHaixin000 committed:
Merge branch 'master' of https://github.com/LiHaixin000/ip

LiHaixin000 committed:
Resolve bug on application cannot run

LiHaixin000 committed:
Use GitHub Actions to set up Continuous Integration

LiHaixin000 committed:
Add extension to allow sorting by deadline and event start date

LiHaixin000 committed:
Improve UI for application interface

LiHaixin000 committed:
Introduce personality-styles to commands

LiHaixin000 committed:
Add more error handling for AddDeadlineCommand, AddEventCommand, AddTodoCommand

LiHaixin000 committed:
Add more error handling for DeleteCommand and FindCommand

LiHaixin000 committed:
Add more error handling for MarkCommand and UnmarkCommand

LiHaixin000 committed:
Merge branch 'branch-A-MoreErrorHandling'

LiHaixin000 committed:
Update JUnit class to support new features

LiHaixin000 committed:
Update text file

LiHaixin000 committed:
Replace all ocurrence of duke with arts

LiHaixin000 committed:
Update user guide

LiHaixin000 committed:
Edit README

LiHaixin000 committed:
Fix bug on storage exception

LiHaixin000 committed:
Fix github issue problems

LiHaixin000 committed:
Fix bug on data file creation

LiHaixin000 committed:
Fix bug on exception handling

LiHaixin000 committed:
Fix test cases

LiHaixin000 committed:
Fix style issue

LiHaixin000 committed:
Edit workflow
